### PR TITLE
Fixed logic in recovery file name detection

### LINKED ISSUE #4893: Recover not working with default options
The recover tests pass, but a test of the feature shows that it does not work as expected when run outside of the test harness. In time_integrators/implicit-euler with the ie.i input file:

###### Adding a checkpoints block to Outputs with default options (no options specified) works as promised, producing 2 rolling checkpoints.

    $ ls ie_checkpoints_cp/
    0002.rd-0     0002_mesh.cpr 0003.xdr.0000 0004.xdr      0005.rd-0     0005_mesh.cpr
    0002.xdr      0003.rd-0     0003_mesh.cpr 0004.xdr.0000 0005.xdr
    0002.xdr.0000 0003.xdr      0004.rd-0     0004_mesh.cpr 0005.xdr.0000

###### Changing num_steps to 10 and running with --recover produces an error:

    $ modules-dbg -i ie.i --recover
    *** ERROR ***
    Unable to find suitable recovery file!

###### Telling MOOSE where to find the checkpoints squelches the error, but the time stepper starts the simulation from the beginning:

    $ modules-dbg -i ie.i --recover=ie_checkpoints_cp/0005
     
    [snipped header]
    
    Time Step  0, time = 0
                    dt = 0

    Postprocessor Values:
    +----------------+----------------+
    | time           | l2_err         |
    +----------------+----------------+
    |   0.000000e+00 |   0.000000e+00 |
    +----------------+----------------+
    
    Time Step  1, time = 0.25
                    dt = 0.25
     0 Nonlinear |R| = 3.062746e+00
     1 Nonlinear |R| = 1.908510e-05
     2 Nonlinear |R| = 1.058361e-10
     Solve Converged!
    
    ...etc

### PR DESCRIPTION
**Pull request by aeslaughter**
(closes #4893)

### CODE REVIEW

File: framework/src/actions/SetupRecoverFileBaseAction.C [hunk 1]
```diff
@@ -134,9 +135,26 @@ SetupRecoverFileBaseAction::getCheckpointFiles(std::set<std::string> & files)
...
+      if (params.isParamValid("file_base"))
+        checkpoint_dirs.insert(common->getParam<std::string>("file_base") + "_cp");
+      else
+      {
+        std::ostringstream oss;
+        oss << "_" << (*it)->getShortName() << "_cp";
+        checkpoint_dirs.insert(FileOutput::getOutputFileBase(_app, oss.str()));
+      }
+    }
+  }
```
Comment (permcody) [inline framework/src/actions/SetupRecoverFileBaseAction.C]: I hate to be critical of both your patches but this a bit clunky too. This patch now loops through a list of unrelated actions inspecting the type to make a decision about something. Let's chat about alternative designs that don't have to interact with the Factory or other objects so much.

Comment (aeslaughter) [inline framework/src/actions/SetupRecoverFileBaseAction.C]: I need to inspect the Output objects, but they do not exist at this point. And, we can't build them as early as this needs to happen. If you have better ideas, that would be great.

**Comment by permcody** [comment]
You don't really need to inspect the Output objects since your patch gets around that. Really you need a way to inspect all of the input file parameters that could impact the recover file base.

One way to accomplish this in the Action system would be to associate your existing Action with the blocks that contain the parameters you need to inspect:

e.g  Associate `SetupRecoverFileBaseAction` with the `Outputs/*` syntax. 

The action system would be several instances of that Action (one for each unique block in the input file) but each would then automatically have access to the parameter that you're trying to get to through by going to the Warehouse and looping over them. Your logic would be very straightforward because you'd know your type and everything else too. Your logic would then be to attempt to retrieve the most current file and set the recovery file directly on the app. No loops, no casting, just a few lines of clean code.

**Comment by permcody** [comment]
One more thing: I think the closest thing we have to this now is the `CopyNodalVarsAction`. It is associated with the `Variables/*` and `AuxVariables/*` blocks. It gets a chance to fire naturally for each new block in the input file to make one decision, much like your case here.

**Comment by aeslaughter** [comment]
I like that idea, I never think of having multiple syntax associates, I
will take a look. Thanks.

**Comment by permcody** [comment]
I highly recommend that you use the Debug/show_actions=true when you build this. You'll see each "instance" of an action created for each section of the input file to satisfy a task (in yellow). Some actions will have the cyan/magenta set of actions listed to the right. Those Actions instances are reused for multiple tasks.

**Comment by aeslaughter** [comment]
@permcody There is a problem with this approach, `registerActionSyntax("SetupRecoverFileBaseAction", "Outputs/*")` works for output objects that have full blocks defined

**Comment by aeslaughter** [comment]
@permcody I do not think this will work, using `registerActionSyntax("SetupRecoverFileBaseActin", "Outputs/*")` fires off the action with each output block, but the action does not contain any valuable information, I need to know the "file_base" parameter with is stored on the "AddOutput" MooseObjectAction. I still will need to probe the warehouse.

**Comment by permcody** [comment]
Well you could fire it for both the parent blocks and the child blocks. Wherever you need to inspect parameters.

**Comment by aeslaughter** [comment]
Decided to leave this as is, the multiple action syntax approach has difficulties as well.

**Pull request merged**